File: core/lexer/lexer.ts
import { TokenType, Token } from "../types/types";

export class Lexer {
  private position = 0;
  private char: string | null = "";

  private peekChar() {
    if (this.position >= this.input.length) {
      return ""; // Return a null or EOF marker if at the end
    } else {
      return this.input[this.position];
    }
  }

  constructor(private input: string) {
    this.readChar(); // Initialize the first character
  }

  private readChar() {
    if (this.position >= this.input.length) {
      this.char = null;
    } else {
      this.char = this.input[this.position];
    }
    this.position++;
  }

  private skipWhitespace() {
    while (this.char && /\s/.test(this.char)) {
      this.readChar();
    }
  }

  public nextToken(): Token {
    this.skipWhitespace();

    // Check for end of file first
    if (this.char === null) {
      return { type: TokenType.EOF, value: "" };
    }

    // Handle symbols and operators
    switch (this.char) {
      case "+":
        this.readChar();
        if ((this.char as string) === "+") {
          this.readChar();
          return { type: TokenType.Incrementation, value: "++" }
        }
        return { type: TokenType.Operator, value: "+" };
      case "-":
        this.readChar();
        if ((this.char as string) === "-") {
          this.readChar();
          return { type: TokenType.Incrementation, value: "--" }
        }
        return { type: TokenType.Operator, value: "-" };
      case "*":
        this.readChar();
        return { type: TokenType.Operator, value: "*" };
      case "/":
        this.readChar();
        return { type: TokenType.Operator, value: "/" };

      case "#":
        const comment = this.readComment();
        return { type: TokenType.Comment, value: comment };
      case ";":
        this.readChar();
        return { type: TokenType.Operator, value: ";" };

      case '"':
        this.readChar(); // Consume the opening quote
        let strValue = "";
        while (this.char !== null && this.char !== '"') {
          if (this.char === "\\") {
            // Handle backslash
            this.readChar();
            // Optional: Add logic to handle \n, \t, etc.
            strValue += this.char;
          } else {
            strValue += this.char;
          }
          this.readChar();
        }
        this.readChar(); // Consume the closing quote
        return { type: TokenType.String, value: strValue };

        case ".":
            this.readChar(); // now this.char is the char after "."
            if (this.isDigit(this.char)) {  // check current char, not peekChar
                const doubleValue = "." + this.readNumber();
                return { type: TokenType.Double, value: doubleValue };
            }
            return { type: TokenType.Dot, value: "." };
      case ">":
        this.readChar();
        if ((this.char as string) === "=") {
          this.readChar();
          return { type: TokenType.Operator, value: ">=" };
        }
        return { type: TokenType.Operator, value: ">" };
      case "<":
        this.readChar();
        if ((this.char as string) === "=") {
          this.readChar();
          return { type: TokenType.Operator, value: "<=" };
        }
        return { type: TokenType.Operator, value: "<" };
      case "!":
        this.readChar();
        if ((this.char as string) === "=") {
          this.readChar();
          return { type: TokenType.Operator, value: "!=" };
        }
        return { type: TokenType.Operator, value: "!" };
      case "=":
        this.readChar();
        if ((this.char as string) === "=") {
          this.readChar();
          return { type: TokenType.Operator, value: "==" };
        }
        return { type: TokenType.Operator, value: "=" };

      case "&":
        this.readChar();
        if ((this.char as string) === "&") {
          this.readChar();
          return { type: TokenType.Operator, value: "&&" };
        }
        return { type: TokenType.Illegal, value: "&" };

      case "|":
        this.readChar();
        if ((this.char as string) === "|") {
          this.readChar();
          return { type: TokenType.Operator, value: "||" };
        }
        return { type: TokenType.Illegal, value: "|" };

      case "%":
        this.readChar();
        return { type: TokenType.Operator, value: "%" };

      case "[":
        this.readChar();
        return { type: TokenType.Operator, value: "[" };
      case "]":
        this.readChar();
        return { type: TokenType.Operator, value: "]" };

      case "(":
        this.readChar();
        return { type: TokenType.Operator, value: "(" };
      case ")":
        this.readChar();
        return { type: TokenType.Operator, value: ")" };

      case "{": 
        this.readChar();
        return { type: TokenType.Operator, value: "{" };
      case "}":
        this.readChar();
        return { type: TokenType.Operator, value: "}" };

      default:
        // Handle words (Keywords and Identifiers)
        if (this.isLetter(this.char)) {
          const literal = this.readIdentifier();
          const keywords = [
            "let",
            "printf",
            "say",
            "if",
            "const",
            "func",
            "True",
            "False",
            "else",
            "while",
            ">",
            "<",
            ">=",
            "<=",
            "==",
            "!=",
            "&&",
            "||",
            "!",
          ];

          // Check if word is a Keyword or Identifier
          const type = keywords.includes(literal)
            ? TokenType.Keyword
            : TokenType.Identifier;
          return { type, value: literal };
        }

        // Handle Numbers
        if (this.isDigit(this.char)) {
            const num = this.readNumber();
            if (num.includes(".")) {
                return { type: TokenType.Double, value: num };
            }
            return { type: TokenType.Number, value: num };
        }

        // If we don't recognize it, it's an illegal character
        const illegalChar = this.char || "";
        this.readChar();
        return { type: TokenType.Illegal, value: illegalChar };
    }
  }

  private isLetter(char: string | null): boolean {
    return !!char && /[a-zA-Z_]/.test(char);
  }

  private isDigit(char: string | null): boolean {
    return !!char && /[0-9]/.test(char);
  }


  private isInteger(char: string | null): boolean {
    return !!char && /[0-9]/.test(char);
  }

  private readIdentifier(): string {
      let literal = "";
      while (this.isLetter(this.char) || (literal.length > 0 && this.isDigit(this.char))) {
          literal += this.char;
          this.readChar();
      }
      return literal;
  }

  private readComment(): string {
    let comment = "";
    while (this.char !== null && this.char !== "\n") {
      comment += this.char;
      this.readChar();
    }
    return comment.trim();
  }

  private readNumber(): string {
    let literal = "";
    while (this.isDigit(this.char) || this.char === ".") {
      literal += this.char;
      this.readChar();
    }
    return literal;
  }
}

File: src/core/lexer/lexer.go
package lexer

import (
	"quill/src/core/types"
	"slices"
	"strings"
	"unicode"
)

type Lexer struct {
	input    []rune
	position int
	char     rune
}

func NewLexer(input string) *Lexer {
	l := &Lexer{input: []rune(input)}
	l.readChar()
	return l
}

func (l *Lexer) Tokenize() []types.Token {
	var tokens []types.Token
	for {
		tok := l.NextToken()
		tokens = append(tokens, tok)
		if tok.Type == types.EOF {
			break
		}
	}
	return tokens
}

func (l *Lexer) peekChar() rune {
	if l.position >= len(l.input) {
		return 0
	}
	return l.input[l.position]
}

func (l *Lexer) readChar() {
	if l.position >= len(l.input) {
		l.char = 0
	} else {
		l.char = l.input[l.position]
	}
	l.position++
}

func (l *Lexer) skipWhitespace() {
	for l.char != 0 && unicode.IsSpace(l.char) {
		l.readChar()
	}
}

func (l *Lexer) readNumber() string {
	position := l.position - 1
	dotSeen := false

	for unicode.IsDigit(l.char) || (!dotSeen && l.char == '.') {
		if l.char == '.' {
			dotSeen = true
		}
		l.readChar()
	}

	return string(l.input[position : l.position-1])
}

func (l *Lexer) readComment() string {
	position := l.position

	for l.char != 0 && l.char != '\n' {
		l.readChar()
	}
	return strings.TrimSpace(string(l.input[position:l.position]))
}

func (l *Lexer) readIdentifier() string {
	position := l.position - 1

	for unicode.IsLetter(l.char) || unicode.IsDigit(l.char) || l.char == '_' {
		l.readChar()
	}
	return string(l.input[position : l.position-1])
}

func (l *Lexer) NextToken() types.Token {
	l.skipWhitespace()

	for l.char == ';' {
		l.readChar()
		l.skipWhitespace()
	}

	if l.char == 0 {
		return types.Token{Type: types.EOF, Value: ""}
	}

	switch l.char {
	case '+':
		l.readChar()
		if l.char == '+' {
			l.readChar()
			return types.Token{Type: types.Incrementation, Value: "++"}
		}
		return types.Token{Type: types.Operator, Value: "+"}

	case '-':
		l.readChar()
		if l.char == '-' {
			l.readChar()
			return types.Token{Type: types.Incrementation, Value: "--"}
		}
		return types.Token{Type: types.Operator, Value: "-"}

	case '*':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "*"}

	case '/':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "/"}

	case '#':
		l.readChar()
		comment := l.readComment()
		return types.Token{Type: types.Comment, Value: comment}

	case ':':
		l.readChar()
		return types.Token{Type: types.Operator, Value: ":"}

	case ',':
		l.readChar()
		return types.Token{Type: types.Operator, Value: ","}

	case '"':
		l.readChar()
		var strValue = ""
		for l.char != 0 && l.char != '"' {
			if l.char == '\\' {
				l.readChar() // consume backslash
				switch l.char {
				case 'n':
					strValue += "\n"
				case 't':
					strValue += "\t"
				case '"':
					strValue += "\""
				case '\\':
					strValue += "\\"
				default:
					strValue += string(l.char)
				}
			} else {
				strValue += string(l.char)
			}
			l.readChar()
		}
		l.readChar() // consume closing "
		return types.Token{Type: types.Str, Value: strValue}

	case '.':
		l.readChar()
		if unicode.IsDigit(l.char) {
			num := l.readNumber()
			return types.Token{Type: types.Float, Value: "." + num}
		}
		return types.Token{Type: types.Dot, Value: "."}

	case '>':
		l.readChar()
		if l.char == '=' {
			l.readChar()
			return types.Token{Type: types.Operator, Value: ">="}
		}
		return types.Token{Type: types.Operator, Value: ">"}

	case '<':
		l.readChar()
		if l.char == '=' {
			l.readChar()
			return types.Token{Type: types.Operator, Value: "<="}
		}
		return types.Token{Type: types.Operator, Value: "<"}

	case '!':
		l.readChar()
		if l.char == '=' {
			l.readChar()
			return types.Token{Type: types.Operator, Value: "!="}
		}
		return types.Token{Type: types.Operator, Value: "!"}

	case '=':
		l.readChar()
		if l.char == '=' {
			l.readChar()
			return types.Token{Type: types.Operator, Value: "=="}
		}
		return types.Token{Type: types.Operator, Value: "="}

	case '&':
		l.readChar()
		if l.char == '&' {
			l.readChar()
			return types.Token{Type: types.Operator, Value: "&&"}
		}
		return types.Token{Type: types.Illegal, Value: "&"}

	case '|':
		l.readChar()
		if l.char == '|' {
			l.readChar()
			return types.Token{Type: types.Operator, Value: "||"}
		}
		return types.Token{Type: types.Illegal, Value: "|"}

	case '%':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "%"}

	case '[':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "["}
	case ']':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "]"}

	case '(':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "("}
	case ')':
		l.readChar()
		return types.Token{Type: types.Operator, Value: ")"}

	case '{':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "{"}
	case '}':
		l.readChar()
		return types.Token{Type: types.Operator, Value: "}"}

	default:
		if unicode.IsLetter(l.char) {
			literal := l.readIdentifier()
			keywords := []string{
				"let", "printf", "say", "if", "const", "func", "true", "false", "else",
				"while", "return", "elif",
			}

			if slices.Contains(keywords, literal) {
				return types.Token{Type: types.Keyword, Value: literal}
			}
			return types.Token{Type: types.Identifier, Value: literal}
		}

		if unicode.IsDigit(l.char) {
			num := l.readNumber()
			if strings.Contains(num, ".") {
				return types.Token{Type: types.Float, Value: num}
			}
			return types.Token{Type: types.Number, Value: num}
		}

		var illegalChar = string(l.char)
		l.readChar()
		return types.Token{Type: types.Illegal, Value: illegalChar}
	}
}
